### PR TITLE
CNV-15994: Add VM snapshot metrics

### PR DESCRIPTION
**Pull request by bgaydosrh**
For 4.11 only.

Jira: https://issues.redhat.com/browse/CNV-15994

Direct doc preview link: http://file.rdu.redhat.com/bgaydos/CNV-15994/virt/logging_events_monitoring/virt-prometheus-queries.html#virt-promql-storage-metrics_virt-prometheus-queries

Tagging Shirly Radko in Jira for Code Review
Tagging Oded Ramraz in Jira to assign a QE Reviewer.

### CODE REVIEW

**Review by akalenyu — COMMENTED** [review]
These metrics originate from https://github.com/kubevirt/kubevirt/pull/7533,
and are more interesting when supplying them with a source VM name and namespace like in the example seen in the PR description.
Could we also document these "parameters"? (vm_name, vm_namespace)
Maybe provide an example?

Regarding the explanation, I would say something more like: 
`Returns the total number of virtual machine disks restored from source VM snapshot`

**Comment by sradco** [comment]
@akalenyu  I believe that the new recording rules are Gauges and not Counters. 
This means that the PR will require a different query then the other metrics here. 
It may also mean that the recording rules names are incorrect with the "_total" suffix.

**Comment by bgaydosrh** [comment]
Hi @sradco - If I am understanding correctly, your last comment indicates that dev changes must be made to correct the rule names and that different (or more?) queries will need to be documented. As this is a late-breaking change in scope, our DPM has deprioritized this story in Jira. 

Please advise if there is information that will enable us to close this doc PR soon. Our feature complete freeze is on Friday (the 22nd).

Thanks
Bob
cc: @ctomasko @apinnick

**Comment by sradco** [comment]
@bgaydosrh Hi, I think with can keep the naming. But each metric should include an example query that shows how it should be used.

**Comment by akalenyu** [comment]
Added second example in https://github.com/openshift/openshift-docs/pull/47429#issuecomment-1188690136

**Comment by bgaydosrh** [comment]
Thanks @akalenyu - I updated the descriptions but still am not able to locate the examples you mention --- the examples on the existing doc page all start with `topk`. Do these new examples not use this format? I looked through the PR https://github.com/kubevirt/kubevirt/pull/7533 and could not find any indication of `topk`. 

In your Jira https://issues.redhat.com/browse/CNV-13691 in the comments, there appear to be what look like examples, but it's confusing to me which version to publish. Keep in mind I do not know this query language.

Can you please provide here 

- The recommended example or examples exactly as the user might enter them
- What each example returns, as do the other examples on https://docs.openshift.com/container-platform/4.10/virt/logging_events_monitoring/virt-prometheus-queries.html

Bob

**Comment by akalenyu** [comment]
Hi Bob

The examples are the ones available in my comment above (https://github.com/openshift/openshift-docs/pull/47429#issuecomment-1188690136), and do not start with `topk`:
* `kubevirt_vmsnapshot_disks_restored_from_source_total{vm_name="simple-vm", vm_namespace="default"}`
`kubevirt_vmsnapshot_disks_restored_from_source_bytes{vm_name="simple-vm", vm_namespace="default"}`

* The return value is the description of the metric (pretty much self-documenting)
So we may get away without explicit return value here?

**Comment by bgaydosrh** [comment]
> Hi Bob
> 
> The examples are the ones available in my comment above ([#47429 (comment)](https://github.com/openshift/openshift-docs/pull/47429#issuecomment-1188690136)), and do not start with `topk`:
> 
> * `kubevirt_vmsnapshot_disks_restored_from_source_total{vm_name="simple-vm", vm_namespace="default"}`
>   `kubevirt_vmsnapshot_disks_restored_from_source_bytes{vm_name="simple-vm", vm_namespace="default"}`
> * The return value is the description of the metric (pretty much self-documenting params too, as we discussed)
>   So we may get away without explicit return value here?

Thanks@akalenyu - I think this should wrap it up. I would agree this should be enough given the parms are self-documenting. Tagging @duyanyan for QE review.

**Comment by akalenyu** [comment]
> > Hi Bob
> > The examples are the ones available in my comment above ([#47429 (comment)](https://github.com/openshift/openshift-docs/pull/47429#issuecomment-1188690136)), and do not start with `topk`:
> > 
> > * `kubevirt_vmsnapshot_disks_restored_from_source_total{vm_name="simple-vm", vm_namespace="default"}`
> >   `kubevirt_vmsnapshot_disks_restored_from_source_bytes{vm_name="simple-vm", vm_namespace="default"}`
> > * The return value is the description of the metric (pretty much self-documenting params too, as we discussed)
> >   So we may get away without explicit return value here?
> 
> Thanks @akalenyu I think this should wrap it up. I would agree this should be enough given the parms are self-documenting. Tagging @duyanyan for QE review.

Awesome! LGTM

**Comment by duyanyan** [comment]
LGTM

File: modules/virt-querying-metrics.adoc [hunk 1]
```diff
@@ -66,12 +66,24 @@ Returns the total amount (in bytes) of the virtual machine's storage-related tra
...
 Returns the total amount of storage writes (in bytes) of the virtual machine's storage-related traffic.
 
-.Example storage-related traffic query
+`kubevirt_vmsnapshot_disks_restored_from_source_total`::
+Returns the total number of virtual machine disks restored from the source virtual machine.
+
+`kubevirt_vmsnapshot_disks_restored_from_source_bytes`::
+Returns the amount of space in bytes restored from the source virtual machine.
+
+.Examples of storage-related traffic queries
```
Comment (ousleyp) [inline modules/virt-querying-metrics.adoc:75]: The [OCP docs contribution guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) state "Do not use more than one command per code block." I think that still applies here, so please split this up into three blocks.

File: modules/virt-querying-metrics.adoc [hunk 2]
```diff
@@ -66,12 +66,24 @@ Returns the total amount (in bytes) of the virtual machine's storage-related tra
 `kubevirt_vmi_storage_write_traffic_bytes_total`::
 Returns the total amount of storage writes (in bytes) of the virtual machine's storage-related traffic.
 
-.Example storage-related traffic query
+`kubevirt_vmsnapshot_disks_restored_from_source_total`::
+Returns the total number of virtual machine disks restored from the source virtual machine.
+
+`kubevirt_vmsnapshot_disks_restored_from_source_bytes`::
+Returns the amount of space in bytes restored from the source virtual machine.
```
Comment (ousleyp) [inline modules/virt-querying-metrics.adoc:73]: Should these have the same heading as the others above them (noting that this section starts with "The following queries can identify VMs that are writing large amounts of data:")? They seem like they're in a different category and could use their own subtitle (maybe something snapshot-related)

**Review by ousleyp — COMMENTED** [review]
I think this needs a little bit of tweaking before I can merge, but the text generally LGTM.

File: modules/virt-querying-metrics.adoc (hunk 1 above)
Comment (bgaydosrh) [reply modules/virt-querying-metrics.adoc:75]: Yep, sounds good.

File: modules/virt-querying-metrics.adoc (hunk 2 above)
Comment (bgaydosrh) [reply modules/virt-querying-metrics.adoc:73]: Yep, they are. I added an ".Examples of storage snapshot data queries" heading as they differ from the traffic-related queires.

File: modules/virt-querying-metrics.adoc [hunk 3]
```diff
@@ -66,13 +66,35 @@ Returns the total amount (in bytes) of the virtual machine's storage-related tra
 `kubevirt_vmi_storage_write_traffic_bytes_total`::
 Returns the total amount of storage writes (in bytes) of the virtual machine's storage-related traffic.
 
-.Example storage-related traffic query
+.Examples of storage-related traffic queries
```
Comment (ousleyp) [inline modules/virt-querying-metrics.adoc:69]: Nit: To align with the other single examples in this file, I think you should revert this line to `.Example storage-related traffic query`

**Review by ousleyp — COMMENTED** [review]
LGTM except for one nit. :)

File: modules/virt-querying-metrics.adoc (hunk 3 above)
Comment (bgaydosrh) [reply modules/virt-querying-metrics.adoc]: Done.

**Pull request merged**